### PR TITLE
Adds -g flag in curl

### LINKED ISSUE #253: cURL Breaking without `-g`
**Describe the bug**
When using GraphQL query bodies in GET requests, the codegen cURL will fail with
`curl: (3) nested brace in URL position 122:`.
However it will pass if `-g` is added to the request.

**To Reproduce**
```
curl --location --request GET 'https://picks-dev.cbssports.com/graphql?query=query%20APIBracketLobbyQuery%20{%0A%20%20bpc:%20gameInstance(game:%20{%20uid:%20%22cbs-ncaab-tournament-challenge%22%20})%20{%0A%20%20%20%20...APIGameInstanceFragment%0A%20%20}%0A%20%20bpm:%20gameInstance(game:%20{%20uid:%20%22cbs-ncaab-tournament-manager%22%20})%20{%0A%20%20%20%20...APIGameInstanceFragment%0A%20%20}%0A%20%20currentUser%20{%0A%20%20%20%20preferredEntryName%0A%20%20%20%20bpcEntries:%20entries(%0A%20%20%20%20%20%20first:%20200%0A%20%20%20%20%20%20gameInstanceUid:%20%22cbs-ncaab-tournament-challenge%22%0A%20%20%20%20)%20{%0A%20%20%20%20%20%20edges%20{%0A%20%20%20%20%20%20%20%20node%20{%0A%20%20%20%20%20%20%20%20%20%20...APILobbyEntryFragment%0A%20%20%20%20%20%20%20%20}%0A%20%20%20%20%20%20}%0A%20%20%20%20}%0A%20%20%20%20bpmEntries:%20entries(%0A%20%20%20%20%20%20first:%20200%0A%20%20%20%20%20%20gameInstanceUid:%20%22cbs-ncaab-tournament-manager%22%0A%20%20%20%20)%20{%0A%20%20%20%20%20%20edges%20{%0A%20%20%20%20%20%20%20%20node%20{%0A%20%20%20%20%20%20%20%20%20%20...APILobbyEntryFragment%0A%20%20%20%20%20%20%20%20}%0A%20%20%20%20%20%20}%0A%20%20%20%20}%0A%20%20}%0A}%0Afragment%20APIGameInstanceFragment%20on%20GameInstance%20{%0A%20%20uid%0A%20%20seasons(first:%20100)%20{%0A%20%20%20%20edges%20{%0A%20%20%20%20%20%20node%20{%0A%20%20%20%20%20%20%20%20id%0A%20%20%20%20%20%20%20%20year%0A%20%20%20%20%20%20%20%20masterProductId%0A%20%20%20%20%20%20%20%20productAbbrev%0A%20%20%20%20%20%20%20%20challengePoolId%0A%20%20%20%20%20%20}%0A%20%20%20%20}%0A%20%20}%0A%20%20currentPeriod%20{%0A%20%20%20%20...APILobbyCurrentPeriodFragment%0A%20%20}%0A}%0Afragment%20APILobbyCurrentPeriodFragment%20on%20Period%20{%0A%20%20locksAt%0A%20%20isPickable%0A%20%20segment%20{%0A%20%20%20%20shouldPromoteToPlay%0A%20%20%20%20season%20{%0A%20%20%20%20%20%20id%0A%20%20%20%20%20%20masterProductId%0A%20%20%20%20%20%20challengePoolId%0A%20%20%20%20}%0A%20%20}%0A}%0Afragment%20APILobbyEntryFragment%20on%20Entry%20{%0A%20%20id%0A%20%20name%0A%20%20fantasyPoints%0A%20%20poolRank%0A%20%20url%0A%20%20needsToMakePicks%0A%20%20avatarUrl%0A%20%20periodPoint%20{%0A%20%20%20%20poolRank%0A%20%20%20%20fantasyPoints%0A%20%20}%0A%20%20pool%20{%0A%20%20%20%20id%0A%20%20%20%20name%0A%20%20%20%20url%0A%20%20%20%20season%20{%0A%20%20%20%20%20%20masterProductId%0A%20%20%20%20}%0A%20%20}%0A}%0A'
```

vs

```
curl -g --location --request GET 'https://picks-dev.cbssports.com/graphql?query=query%20APIBracketLobbyQuery%20{%0A%20%20bpc:%20gameInstance(game:%20{%20uid:%20%22cbs-ncaab-tournament-challenge%22%20})%20{%0A%20%20%20%20...APIGameInstanceFragment%0A%20%20}%0A%20%20bpm:%20gameInstance(game:%20{%20uid:%20%22cbs-ncaab-tournament-manager%22%20})%20{%0A%20%20%20%20...APIGameInstanceFragment%0A%20%20}%0A%20%20currentUser%20{%0A%20%20%20%20preferredEntryName%0A%20%20%20%20bpcEntries:%20entries(%0A%20%20%20%20%20%20first:%20200%0A%20%20%20%20%20%20gameInstanceUid:%20%22cbs-ncaab-tournament-challenge%22%0A%20%20%20%20)%20{%0A%20%20%20%20%20%20edges%20{%0A%20%20%20%20%20%20%20%20node%20{%0A%20%20%20%20%20%20%20%20%20%20...APILobbyEntryFragment%0A%20%20%20%20%20%20%20%20}%0A%20%20%20%20%20%20}%0A%20%20%20%20}%0A%20%20%20%20bpmEntries:%20entries(%0A%20%20%20%20%20%20first:%20200%0A%20%20%20%20%20%20gameInstanceUid:%20%22cbs-ncaab-tournament-manager%22%0A%20%20%20%20)%20{%0A%20%20%20%20%20%20edges%20{%0A%20%20%20%20%20%20%20%20node%20{%0A%20%20%20%20%20%20%20%20%20%20...APILobbyEntryFragment%0A%20%20%20%20%20%20%20%20}%0A%20%20%20%20%20%20}%0A%20%20%20%20}%0A%20%20}%0A}%0Afragment%20APIGameInstanceFragment%20on%20GameInstance%20{%0A%20%20uid%0A%20%20seasons(first:%20100)%20{%0A%20%20%20%20edges%20{%0A%20%20%20%20%20%20node%20{%0A%20%20%20%20%20%20%20%20id%0A%20%20%20%20%20%20%20%20year%0A%20%20%20%20%20%20%20%20masterProductId%0A%20%20%20%20%20%20%20%20productAbbrev%0A%20%20%20%20%20%20%20%20challengePoolId%0A%20%20%20%20%20%20}%0A%20%20%20%20}%0A%20%20}%0A%20%20currentPeriod%20{%0A%20%20%20%20...APILobbyCurrentPeriodFragment%0A%20%20}%0A}%0Afragment%20APILobbyCurrentPeriodFragment%20on%20Period%20{%0A%20%20locksAt%0A%20%20isPickable%0A%20%20segment%20{%0A%20%20%20%20shouldPromoteToPlay%0A%20%20%20%20season%20{%0A%20%20%20%20%20%20id%0A%20%20%20%20%20%20masterProductId%0A%20%20%20%20%20%20challengePoolId%0A%20%20%20%20}%0A%20%20}%0A}%0Afragment%20APILobbyEntryFragment%20on%20Entry%20{%0A%20%20id%0A%20%20name%0A%20%20fantasyPoints%0A%20%20poolRank%0A%20%20url%0A%20%20needsToMakePicks%0A%20%20avatarUrl%0A%20%20periodPoint%20{%0A%20%20%20%20poolRank%0A%20%20%20%20fantasyPoints%0A%20%20}%0A%20%20pool%20{%0A%20%20%20%20id%0A%20%20%20%20name%0A%20%20%20%20url%0A%20%20%20%20season%20{%0A%20%20%20%20%20%20masterProductId%0A%20%20%20%20}%0A%20%20}%0A}%0A'
```

**Expected code snippet and corresponding request**
Add `-g` to all cURL requests by default **or** properly deal with double vs single quotes (https://github.com/postmanlabs/postman-code-generators/pull/251)

**Additional context**
collection: `11254382-24f763ba-01f6-4782-85c5-f65f3809457b`
Version of postman-code-generators/Postman app: Version 7.24.0 (7.24.0)

### PR DESCRIPTION
**Pull request by someshkoli**
fixes #253 
Since the curl snippets fail if the url parameter has nested braces {}[], we need an additional flag  to process the curl command.
Solutions: Add -g flag in curl when request url has nested braces in it.

This PR adds the following changes:
 - checks if the url parameter has nested {}[]
 - Adds regex for checking the same pattern
 - Adds -g flag when nested {}[] is present
 - Accounts the following cases
   - {{}}
   - [[]]
   - {[]}
   - [{}]
 - Adds unit test for this condition

### CODE REVIEW

File: codegens/curl/test/unit/convert.test.js [hunk 1]
```diff
@@ -59,6 +59,45 @@ describe('curl convert function', function () {
...
+        }
+      });
+      options = {
+        longFormat: false
+      };
+      convert(request, options, function (error, snippet) {
+        if (error) {
+          expect.fail(null, null, error);
+        }
+        expect(snippet).to.include("-g"); // eslint-disable-line quotes
```
Comment (umeshp7) [inline codegens/curl/test/unit/convert.test.js:97]: Why can't we use single quotes here?

File: codegens/curl/test/unit/convert.test.js [hunk 2]
```diff
@@ -59,6 +59,45 @@ describe('curl convert function', function () {
...
+          'query': [
+            {
+              'key': 'a',
+              'value': '{{x:[[{['
+            }
+          ]
+        }
+      });
+      options = {
+        longFormat: false
```
Comment (umeshp7) [inline codegens/curl/test/unit/convert.test.js:91]: Why do we need longFormat as `false`?

File: codegens/curl/test/unit/convert.test.js [hunk 3]
```diff
@@ -59,6 +59,45 @@ describe('curl convert function', function () {
...
+            'key': 'foo',
+            'value': '"bar"'
+          }
+        ],
+        'body': {
+          'mode': 'raw',
+          'raw': ''
+        },
+        'url': {
+          'raw': 'https://google.com?a={{x:[[{[',
```
Comment (umeshp7) [inline codegens/curl/test/unit/convert.test.js:76]: Can you split these into different cases for each pattern you are checking?

File: codegens/curl/lib/index.js [hunk 4]
```diff
@@ -31,14 +32,17 @@ self = module.exports = {
     if (timeout > 0) {
       snippet += ` ${form('-m', format)} ${timeout}`;
     }
+    // eslint-disable-next-line no-useless-escape
+    if ((url.match(/[\{\[]/g) || []).length > 1) {
```
Comment (umeshp7) [inline codegens/curl/lib/index.js:36]: Remove unnecessary escape characters. `(/[{[]/g)` <- this should work the same way.

Comment (someshkoli) [reply codegens/curl/lib/index.js:36]: No actually they are not part of a string. Those escape characters are for the regex

Comment (umeshp7) [reply codegens/curl/lib/index.js:36]: Those are redundant escape characters, not required for the regex. Removing them the regex will work the same way.

Comment (umeshp7) [reply codegens/curl/lib/index.js:36]: `// eslint-disable-next-line no-useless-escape` 
Avoid disabling any lint rules as far as possible.

Comment (someshkoli) [reply codegens/curl/lib/index.js:36]: > Those are redundant escape characters, not required for the regex. Removing them the regex will work the same way.

yeah they work without escape character too. Thanks

File: codegens/curl/test/unit/convert.test.js (hunk 3 above)
Comment (someshkoli) [reply codegens/curl/test/unit/convert.test.js:76]: okay

File: codegens/curl/test/unit/convert.test.js (hunk 1 above)
Comment (someshkoli) [reply codegens/curl/test/unit/convert.test.js:97]: We can, picked up  test snippet from existing one and forgot to check.

**Comment by webholik** [comment]
`curl` gives special meaning to `{` and `[` in urls. So even if the request does not contain nested braces,`curl` will change the meaning of the request. For example:

    curl 'http://example.com/?q={foo:bar}' 

will be sent as:

     `http://example.com/?q=foo:bar`.

And all requests with unmatched `{` and `[` will fail.  I would recommend adding the `-g` options by default, since Postman requests can contain `{` and `[` without any special meaning attatched.

**Comment by umeshp7** [comment]
@webholik The changes look okay. 
Could there be a use case where the user would want the braces to be treated differently and have the cURL request fail?

**Comment by webholik** [comment]
> Could there be a use case where the user would want the braces to be treated differently and have the cURL request fail?

The requests are coming from Postman and the only use of braces there is to fill in variables. The variables will be substituted before the request going to codegen. And since we are using the `-g` flag, the presence or absence of braces will not affect cURL.

**Review by umeshp7 — APPROVED** [review]

**Pull request merged**